### DISCUSSION TITLE
Where and how to create entity model which will be connected with identity model class?

### DISCUSSION THREAD

**Discussion by scarybook** [fullstackhero/blazor-starter-kit#373] [Q&A]
In reference to current project architecture where should I create new entity model when I want to create sth like Department Entity? Each of department will have list of Users ( BlazorHeroUser class ) so creating department model class in Domain project will cause circular dependency for Infrastructure project.

Any suggestions how to deal with it?

Maybe all models or just BlazorHeroUser from Infrastructure project should be moved to other project which doesnt have dependency from Application Domain.

What do you think about that?

**Comment by Zigimun** [comment]
I would realy like to know that as well. So I can get the list of departments based on user and show the viable departments to active user if he is in the list of Users for in the department.

Please help :)

Mybe we should move the Identity models to the domain project. Where we can reference it then by other Entities. But that might be the wrong way of doing it. But the problem is the reference of the Interfaces

public class BlazorHeroUser : IdentityUser<string>, **IChatUser, IAuditableEntity<string>**


I have found the way to do this. Please someone confirm this is the correct way
  
**So in Domain I create 2 entities**
```
    public class Sector : AuditableEntity<int>
    {
        public string Name { get; set; }
        public virtual ICollection<Process> Processes { get; set; }
    }
```

```
    public class Process : AuditableEntity<int>
    {
        public string Name { get; set; }
        public int SectorId { get; set; }
        public virtual Sector Sector { get; set; }
    }
```

**In Application I create** 

```
    public interface IProcessUser
    {
        public string FirstName { get; set; }

        public string LastName { get; set; }

        [Column(TypeName = "text")]
        public string ProfilePictureDataUrl { get; set; }
    }
```
```
    public interface IProcess<TUser> where TUser : IProcessUser
    {
        public int Id { get; set; }
        public string Name { get; set; }
        public ICollection<TUser> Users { get; set; }
    }
```
```
    public class Process<TUser> : Process, IProcess<TUser> where TUser : IProcessUser
    {
        public ICollection<TUser> Users { get; set; }
    }
```

**In Infrastructure I do this**
BlazorHeroContext
```
        public DbSet<Sector> Sectors { get; set; }
        public DbSet<Process<BlazorHeroUser>> ProcessUsers { get; set; }
```

BlazorHeroUser I add this

```
    public class BlazorHeroUser : IdentityUser<string>, IChatUser , IProcessUser, IAuditableEntity<string>
     {
               public virtual ICollection<Process<BlazorHeroUser>> ProcessUsers { get; set; }
     }
```

**Reply by Zigimun** [reply]
This is the answer I think and it should be part of the template so other users of the template see how this could be created, But there is problem like in chat that some of the queries are in Application Project and some queries are in Infrastructure Project, this is what realy bothers me.

Plus the problem is I cannot access that table (ProcessUsers) via dbcontext.

It would be very nice if some more senior dev. helps here with this problem since this is real pain in the ***.

I was considering creating another table of db.Users where I would store the User Id and some Guid whenever I register the user in the system.

**Reply by scarybook** [reply]
I think that making binding class in domain project will be simplier for this moment. sth like what you said - another entity table for user which will store userId from Blazorherouser (guid) and all needed references from domain models. quite a bit messy workaround but blazorherouser will be extented by only one property (userExt) from domain project.

**Reply by Zigimun** [reply]
You probably mean like I said on my last line of comment? Its a big workaround and pretty nasty since 20% of my domains will need `List<User>` or `User`. For the queries. Since the product I am making the customers wants restrictons based on users for each process. And it would be so easier to bind it via Identity

**Reply by Zigimun** [reply]
Could you provide vode how would you do it?

**Reply by scarybook** [reply]
So how if we create few classes in Domain, one of them as below:
```
public class UserExt : AuditableEntity<int>
{
    public IEnumerable<UserExt> Deputies { get; set; }
    public IEnumerable<Absence> Absences { get; set; }

    public Department Department { get; set; }
    public int DepartmentId { get; set; }

    public string BlazorHeroUserId { get; set; } // a property to store userId from identity user model
}
```

in deparment we can easy define collection of UserExt.
then we only add one property in BlazorHeroUser class:

```
public class BlazorHeroUser : IdentityUser<string>, IChatUser, IAuditableEntity<string>
    {
        //some code...
        public UserExt UserExt { get; set; }
        //some code...
    }
```
    
 at the end made few changes to create UserExt binded together with BlazorHeroUser when new User is created. the main thing is that we will be operating mostly on UserExt in our app but if we need identity user data we can simply get it by its ID from UserExt

what will you say about that approach?

**Reply by Zigimun** [reply]
I think this is the correct approach. But I would store the email as well, if you use the emailsender service that you do not need to go and fetch the data from Identity.

But one of more expirienced programmers should confirm this approach as the correct one. I bet there is a way with interfaces and so on, but this I think is more SOLID and much easier to use. Plus its part of the domain data and should be there.

And if you ask me one kind of this kind of situation should be in the template. Like on product should be list of users, so that we can query and see how many users bought the product.... Mybe this is lame example but something like this.

**Reply by sergeytys** [reply]
Wouldn't it be easier to refactor and move the BlazorHeroUser to the Domain project straight? instead of creating ext class on top of it? - what logic in other part suffers from such a plain move? any ideas?
